Add stubs for Trusted Types API

lukewarlow opened the pull request:
#### 74d050dabf126cdca0dd92b7cf303dd5dc925f63
<pre>
Add stubs for Trusted Types API
<a href="https://bugs.webkit.org/show_bug.cgi?id=266980">https://bugs.webkit.org/show_bug.cgi?id=266980</a>

Reviewed by Ryosuke Niwa and Alex Christensen.

This patch adds a runtime feature flag and the initial bindings for the Trusted Types API.
None of the enforcement logic is implemented yet, nor any changes to the DOM sinks.

Spec: <a href="https://w3c.github.io/trusted-types/dist/spec">https://w3c.github.io/trusted-types/dist/spec</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/securitypolicyviolation/source-file-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/DOMParser-parseFromString-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/DOMWindowTimers-setTimeout-setInterval-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Document-write-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Element-insertAdjacentHTML-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Element-insertAdjacentText-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Element-outerHTML-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Element-setAttribute-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Element-setAttributeNS-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/GlobalEventHandlers-onclick-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLScriptElement-internal-slot-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Node-multiple-arguments-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Range-createContextualFragment-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedType-AttributeNodes-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicy-CSP-no-name-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicy-CSP-wildcard-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicy-createXXX-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-blocking-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-constants-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-createXYZTests-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-cspTests-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-cspTests-noNamesGiven-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-cspTests-none-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-cspTests-none-skip-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-cspTests-wildcard-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-nameTests-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-unenforced-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-defaultPolicy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getAttributeType-namespace-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getPropertyType-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-isXXX-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-metadata-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Window-TrustedTypes-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/WorkerGlobalScope-eval-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/WorkerGlobalScope-importScripts-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-Document-execCommand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-Node-multiple-arguments-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-DOMParser-parseFromString-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-DOMWindowTimers-setTimeout-setInterval-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Document-write-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Element-insertAdjacentHTML-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Element-outerHTML-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Element-setAttribute-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Element-setAttributeNS-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-HTMLElement-generic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Range-createContextualFragment-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-text-node-insertion-into-script-element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/csp-block-eval-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/default-policy-callback-arguments-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/default-policy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/default-policy-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/empty-default-policy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/empty-default-policy-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-no-tt-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-tt-default-policy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-tt-no-default-policy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-function-constructor-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-no-csp-no-tt-default-policy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-no-csp-no-tt-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-with-permissive-csp-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/no-require-trusted-types-for-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/no-require-trusted-types-for-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/require-trusted-types-for-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-createHTMLDocument-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-duplicate-names-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-duplicate-names-list-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-duplicate-names-list-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-duplicate-names-without-enforcement-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-eval-reporting-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-eval-reporting-no-unsafe-eval-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-eval-reporting-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-from-literal-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-reporting-check-report-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-source-file-path-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-svg-script-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-tojson-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/tt-block-eval-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/worker-constructor.https-expected.txt:
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/CreateHTMLCallback.h: Added.
* Source/WebCore/dom/CreateHTMLCallback.idl: Added.
* Source/WebCore/dom/CreateScriptCallback.h: Added.
* Source/WebCore/dom/CreateScriptCallback.idl: Added.
* Source/WebCore/dom/CreateScriptURLCallback.h: Added.
* Source/WebCore/dom/CreateScriptURLCallback.idl: Added.
* Source/WebCore/dom/TrustedHTML.cpp: Added.
(WebCore::TrustedHTML::create):
(WebCore::TrustedHTML::TrustedHTML):
* Source/WebCore/dom/TrustedHTML.h: Added.
(WebCore::TrustedHTML::toString const):
(WebCore::TrustedHTML::toJSON const):
* Source/WebCore/dom/TrustedHTML.idl: Added.
* Source/WebCore/dom/TrustedScript.cpp: Added.
(WebCore::TrustedScript::create):
(WebCore::TrustedScript::TrustedScript):
* Source/WebCore/dom/TrustedScript.h: Added.
* Source/WebCore/dom/TrustedScript.idl: Added.
* Source/WebCore/dom/TrustedScriptURL.cpp: Added.
(WebCore::TrustedScriptURL::create):
(WebCore::TrustedScriptURL::TrustedScriptURL):
* Source/WebCore/dom/TrustedScriptURL.h: Added.
(WebCore::TrustedScriptURL::toString const):
(WebCore::TrustedScriptURL::toJSON const):
* Source/WebCore/dom/TrustedScriptURL.idl: Added.
* Source/WebCore/dom/TrustedTypePolicy.cpp: Added.
(WebCore::TrustedTypePolicy::create):
(WebCore::TrustedTypePolicy::TrustedTypePolicy):
(WebCore::TrustedTypePolicy::createHTML):
(WebCore::TrustedTypePolicy::createScript):
(WebCore::TrustedTypePolicy::createScriptURL):
* Source/WebCore/dom/TrustedTypePolicy.h: Added.
(WebCore::TrustedTypePolicy::name const):
* Source/WebCore/dom/TrustedTypePolicy.idl: Added.
* Source/WebCore/dom/TrustedTypePolicyFactory.cpp: Added.
(WebCore::TrustedTypePolicyFactory::create):
(WebCore::TrustedTypePolicyFactory::createPolicy):
(WebCore::TrustedTypePolicyFactory::isHTML const):
(WebCore::TrustedTypePolicyFactory::isScript const):
(WebCore::TrustedTypePolicyFactory::isScriptURL const):
(WebCore::TrustedTypePolicyFactory::emptyHTML const):
(WebCore::TrustedTypePolicyFactory::emptyScript const):
(WebCore::TrustedTypePolicyFactory::getAttributeType const):
(WebCore::TrustedTypePolicyFactory::getPropertyType const):
* Source/WebCore/dom/TrustedTypePolicyFactory.h: Added.
(WebCore::TrustedTypePolicyFactory::defaultPolicy const):
* Source/WebCore/dom/TrustedTypePolicyFactory.idl: Added.
* Source/WebCore/dom/TrustedTypePolicyOptions.h: Added.
* Source/WebCore/dom/TrustedTypePolicyOptions.idl: Added.
* Source/WebCore/dom/WindowOrWorkerGlobalScope+TrustedTypes.idl: Added.
* Source/WebCore/dom/WindowOrWorkerGlobalScopeTrustedTypes.cpp: Added.
(WebCore::DOMWindowTrustedTypes::supplementName):
(WebCore::WorkerGlobalScopeTrustedTypes::supplementName):
(WebCore::DOMWindowTrustedTypes::DOMWindowTrustedTypes):
(WebCore::DOMWindowTrustedTypes::from):
(WebCore::DOMWindowTrustedTypes::trustedTypes const):
(WebCore::WorkerGlobalScopeTrustedTypes::WorkerGlobalScopeTrustedTypes):
(WebCore::WorkerGlobalScopeTrustedTypes::from):
(WebCore::WorkerGlobalScopeTrustedTypes::trustedTypes const):
(WebCore::WindowOrWorkerGlobalScopeTrustedTypes::trustedTypes):
* Source/WebCore/dom/WindowOrWorkerGlobalScopeTrustedTypes.h: Added.
* Source/WebCore/page/ios/DOMTimerHoldingTank.h:
* Source/WebCore/Modules/WebGPU/GPUOutOfMemoryError.h:

Canonical link: <a href="https://commits.webkit.org/273817@main">https://commits.webkit.org/273817@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e807cbd6b9fb2af0751d42c567e7045a1458437c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36699 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15638 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38932 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39346 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32898 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37943 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18115 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12749 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31483 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37262 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13177 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32446 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11542 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11546 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32711 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40593 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/30798 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33241 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33062 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37467 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/36461 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11819 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9645 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35587 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13489 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/43259 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8332 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12224 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/8948 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12706 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/dom/TrustedHTML.idl:
```diff
@@ -0,0 +1,9 @@
+interface TrustedHTML {
+  stringifier;
+  DOMString toJSON();
+  static TrustedHTML fromLiteral(object templateStringsArray);
```
nit: 4-space indent

nt1m commented on Source/WebCore/dom/TrustedScript.idl:
```diff
@@ -0,0 +1,10 @@
+interface TrustedScript {
+  stringifier;
+  DOMString toJSON();
+  static TrustedScript fromLiteral(object templateStringsArray);
```
nit: 4-space indent

nt1m commented on Source/WebCore/dom/TrustedScriptURL.idl:
```diff
@@ -0,0 +1,10 @@
+interface TrustedScriptURL {
+  stringifier;
+  DOMString toJSON();
+  static TrustedScriptURL fromLiteral(object templateStringsArray);
```
nit: 4-space indent

nt1m commented on Source/WebCore/dom/TrustedTypePolicyFactory.idl:
```diff
@@ -0,0 +1,23 @@
+        DOMString tagName,
+        DOMString property,
+        optional DOMString elementNs = "");
+    readonly attribute TrustedTypePolicy? defaultPolicy;
```
nit: Can you add a blank line before / after each multi-line item? Makes it easier to read.

nt1m commented on Source/WebCore/dom/TrustedTypePolicyOptions.idl:
```diff
@@ -0,0 +1,5 @@
+dictionary TrustedTypePolicyOptions {
+  CreateHTMLCallback createHTML;
+  CreateScriptCallback createScript;
+  CreateScriptURLCallback createScriptURL;
```
nit: 4 space indent

nt1m commented on Source/WebCore/dom/TrustedTypePolicyFactory.cpp:
```diff
@@ -0,0 +1,107 @@
+{
+    if (JSC::jsDynamicCast<JSTrustedHTML*>(value))
+        return true;
+    return false;
```
```suggestion
    return JSC::jsDynamicCast<JSTrustedHTML*>(value);
```

Here and below, unless you're planning to add extra logic, this is better.

nt1m commented on Source/WebCore/dom/TrustedTypePolicy.cpp:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    return Exception { ExceptionCode::TypeError, "Policy " + m_name +
+        "'s TrustedTypePolicyOptions did not specify a 'createHTML' member." };
```
```suggestion
    return Exception {
        ExceptionCode::TypeError,
        makeString("Policy "_s, m_name, 
        "'s TrustedTypePolicyOptions did not specify a 'createHTML' member."_s)
    };
```
nit: use makeString here

nt1m commented on Source/WebCore/dom/TrustedTypePolicy.cpp:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    return Exception { ExceptionCode::TypeError, "Policy " + m_name +
+        "'s TrustedTypePolicyOptions did not specify a 'createScript' member." };
```
ditto

nt1m commented on Source/WebCore/dom/TrustedTypePolicy.cpp:
```diff
@@ -0,0 +1,106 @@
+            return Exception { ExceptionCode::ExistingExceptionError };
+    }
+
+    return Exception { ExceptionCode::TypeError, "Policy " + m_name +
```
ditto

nt1m commented on Source/WebCore/dom/CreateHTMLCallback.idl:
I missed this earlier, but IDL files should also have copyright headers.

rwlbuis commented on Source/WebCore/dom/TrustedTypePolicyFactory.h:
```diff
@@ -0,0 +1,66 @@
+    Ref<TrustedTypePolicy> createPolicy(const String& policyName, const TrustedTypePolicyOptions&);
+    bool isHTML(JSC::JSValue);
+    bool isScript(JSC::JSValue);
+    bool isScriptURL(JSC::JSValue);
```
Could these be const?

lukewarlow replied:
Good catch

rwlbuis commented on Source/WebCore/dom/TrustedTypePolicyFactory.h:
```diff
@@ -0,0 +1,66 @@
+    Ref<TrustedScript> emptyScript() const;
+
+    String getAttributeType(const String& tagName, const String& attribute, const String& elementNs, const String& attrNs);
+    String getPropertyType(const String& tagName, const String& property, const String& elementNs);
```
Ditto?

rwlbuis commented on Source/WebCore/Modules/WebGPU/GPUOutOfMemoryError.h:
```diff
@@ -28,6 +28,7 @@
 #include "WebGPUOutOfMemoryError.h"
 #include <wtf/RefCounted.h>
 #include <wtf/RefPtr.h>
+#include <wtf/text/WTFString.h>
```
Is this change needed for unified builds or something like that?

lukewarlow replied:
Yeah it doesn't compile otherwise.

rwlbuis commented on Source/WebCore/page/ios/DOMTimerHoldingTank.h:
```diff
@@ -29,6 +29,7 @@
 
 #include "Timer.h"
 #include <wtf/Forward.h>
+#include <wtf/WeakHashSet.h>
```
Ditto?

lukewarlow replied:
Yep same as above it doesn't compile otherwise.

rwlbuis commented on Source/WebCore/dom/WindowOrWorkerGlobalScopeTrustedTypes.cpp:
```diff
@@ -0,0 +1,133 @@
+
+// WindowOrWorkerGlobalScopeTrustedTypes.
+
+ExceptionOr<TrustedTypePolicyFactory*> WindowOrWorkerGlobalScopeTrustedTypes::trustedTypes(ScriptExecutionContext&, LocalDOMWindow& window)
```
Can an exception really happen?

lukewarlow replied:
Ah good catch no I don't think it can.

rwlbuis commented on Source/WebCore/dom/WindowOrWorkerGlobalScopeTrustedTypes.cpp:
```diff
@@ -0,0 +1,133 @@
+    return DOMWindowTrustedTypes::from(window)->trustedTypes();
+}
+
+TrustedTypePolicyFactory* WindowOrWorkerGlobalScopeTrustedTypes::trustedTypes(ScriptExecutionContext&, WorkerGlobalScope& scope)
```
And no exception can happen here?

rwlbuis commented on Source/WebCore/dom/WindowOrWorkerGlobalScopeTrustedTypes.cpp:
```diff
@@ -0,0 +1,133 @@
+
+namespace WebCore {
+
+class DOMWindowTrustedTypes : public Supplement<LocalDOMWindow>, public LocalDOMWindowProperty {
```
Personally I would find it easier to read to give DOMWindowTrustedTypes declaration and implementation first, and then same for WorkerGlobalScopeTrustedTypes.

lukewarlow replied:
Changed them round and fixed the return type issue

achristensen07 commented on Source/WebCore/dom/TrustedHTML.cpp:
```diff
@@ -0,0 +1,46 @@
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(TrustedHTML);
+
+// static
```
I don't think this comment is useful.  We have hundreds of create methods in WebKit and most of them are static.

rniwa replied:
Ditto about removing this comment.

achristensen07 commented on Source/WebCore/dom/WindowOrWorkerGlobalScopeTrustedTypes.cpp:
```diff
@@ -0,0 +1,130 @@
+    return DOMWindowTrustedTypes::from(window)->trustedTypes();
+}
+
+// WorkerGlobalScope supplement.
```
Same.  This doesn't add anything that Supplement<WorkerGlobalScope> on the next line doesn't already tell us.

achristensen07 commented on Source/WebCore/dom/WindowOrWorkerGlobalScopeTrustedTypes.cpp:
```diff
@@ -0,0 +1,130 @@
+    mutable RefPtr<TrustedTypePolicyFactory> m_trustedTypes;
+};
+
+
```
nit: extra newline.

rniwa commented on Source/WebCore/dom/WindowOrWorkerGlobalScopeTrustedTypes.cpp:
```diff
@@ -0,0 +1,130 @@
+
+// DOMWindowTrustedTypes supplement.
+class DOMWindowTrustedTypes : public Supplement<LocalDOMWindow>, public LocalDOMWindowProperty {
+    WTF_MAKE_FAST_ALLOCATED;
```
Please make this ISO heaped.

lukewarlow replied:
The other Suplements are all fast allocated and when I try to use ISO I get compiler errors about missing delete operations?

rniwa commented on Source/WebCore/dom/WindowOrWorkerGlobalScopeTrustedTypes.cpp:
```diff
@@ -0,0 +1,130 @@
+
+namespace WebCore {
+
+// DOMWindowTrustedTypes supplement.
```
This comment seems unnecessary.

rniwa commented on Source/WebCore/dom/TrustedTypePolicyFactory.idl:
```diff
@@ -0,0 +1,50 @@
+    readonly attribute TrustedScript emptyScript;
+
+    DOMString? getAttributeType(
+        DOMString tagName,
```
Please put this all in a single line.
There is no reason we need to line-break so many times here.

rniwa commented on Source/WebCore/dom/TrustedTypePolicyFactory.idl:
```diff
@@ -0,0 +1,50 @@
+        optional DOMString elementNs = "",
+        optional DOMString attrNs = "");
+
+    DOMString? getPropertyType(
```
Ditto.

rniwa commented on Source/WebCore/dom/TrustedTypePolicyFactory.idl:
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (C) 2024 Igalia S.L. All rights reserved.
```
Are we sure we're not copying any Blink files here?
If so, copyright notice does need to mention Chromium authors.

lukewarlow replied:
I'm fairly confident these files are all either IDL copied from the spec or written from scratch.

rniwa commented on Source/WebCore/dom/TrustedTypePolicyFactory.h:
```diff
@@ -0,0 +1,64 @@
+    String getPropertyType(const String& tagName, const String& property, const String& elementNs) const;
+
+    RefPtr<TrustedTypePolicy> defaultPolicy() const { return m_defaultPolicy; }
+private:
```
Nit: Need a blank line before `private:`.

rniwa commented on Source/WebCore/dom/TrustedTypePolicyFactory.cpp:
```diff
@@ -0,0 +1,101 @@
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(TrustedTypePolicyFactory);
+
+// static
```
This comment seems unnecessary.

rniwa commented on Source/WebCore/dom/TrustedTypePolicy.cpp:
```diff
@@ -0,0 +1,115 @@
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(TrustedTypePolicy);
+
+// static
```
This comment seems unnecessary.

rniwa commented on Source/WebCore/dom/TrustedTypePolicy.h:
```diff
@@ -0,0 +1,62 @@
+    TrustedTypePolicy(const String&, const TrustedTypePolicyOptions&);
+
+    String m_name;
+    RefPtr<CreateHTMLCallback> m_createHTMLCallback;
```
Who keeps these callbacks alive?

rniwa commented on Source/WebCore/dom/TrustedScriptURL.cpp:
```diff
@@ -0,0 +1,46 @@
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(TrustedScriptURL);
+
+// static
```
These //static comments are all unnecessary. Please remove them all.

rniwa commented on Source/WebCore/dom/TrustedScript.idl:
```diff
@@ -0,0 +1,34 @@
+
+[
+    EnabledBySetting=TrustedTypesEnabled,
+    Exposed=(Window,Worker)
```
Nit: Need a space after ",".

lukewarlow replied:
This doesn't compile with spaces in the Exposed attribute value, the preprocess-idls.pl file errors. The code just splits on `,` and then does an exact match for values.

rniwa commented on Source/WebCore/dom/TrustedScript.h:
```diff
@@ -0,0 +1,46 @@
+
+    String toString() const { return m_data; }
+    String toJSON() const { return toString(); }
+private:
```
Nit: Need a blank line before "private:"

rniwa commented on Source/WebCore/dom/TrustedHTML.idl:
```diff
@@ -0,0 +1,33 @@
+
+[
+    EnabledBySetting=TrustedTypesEnabled,
+    Exposed=(Window,Worker)
```
Ditto about needing a space after ,

rniwa commented on Source/WebCore/dom/TrustedScript.cpp:
```diff
@@ -0,0 +1,46 @@
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(TrustedScript);
+
+// static
```
Ditto about removing these comments.

rniwa commented on Source/WebCore/dom/CreateHTMLCallback.h:
```diff
@@ -0,0 +1,42 @@
+
+namespace WebCore {
+
+class CreateHTMLCallback : public RefCounted<CreateHTMLCallback>, public ActiveDOMCallback {
```
It's unclear to me how these callbacks are kept alive.
What's the GC model here?

lukewarlow replied:
The idea is that the policy holds a pointer to them which keeps them alive as long as the policy itself is kept alive. Once a policy goes out of scope in JS the callbacks should also be collected.

rniwa replied:
Okay, this PR doesn't implement that GC behavior. Are you planning to implement it in a follow up?

lukewarlow replied:
Just so I fully understand is the issue with the current code that these pointers aren't correctly disposed when the script wrapper for the policy is disposed? Is this something I need to do with the ActiveDOMCallback type or am I missing something more fundamental?

Any changes that need to be made I'm happy to do as follow ups to keep this change as small as possible, or as part of this initial PR if that's prefered.

lukewarlow replied:
Aside from the GC side of things is the PR ready to go? @rniwa

rniwa replied:
We need to make use of either [visit children](https://docs.webkit.org/Deep%20Dive/Architecture/JSWrappers.html#visit-children) or [opaque roots](https://docs.webkit.org/Deep%20Dive/Architecture/JSWrappers.html#opaque-roots).

See MutationObserver / MutationCallback / IntersectionObserver / IntersectionObserverCallback.

rniwa replied:
Yeah, the PR looks good otherwise.

lukewarlow replied:
Okay thanks for that I'll make an issue to address it and do that as a follow up.

rniwa reviewed (changes requested):
Please fix nits & clarify callback lifecycle.

rniwa commented on Source/WebCore/dom/TrustedTypePolicyFactory.idl:
```diff
@@ -0,0 +1,43 @@
+    readonly attribute TrustedHTML emptyHTML;
+    readonly attribute TrustedScript emptyScript;
+
+    DOMString? getAttributeType(DOMString tagName, DOMString attribute, optional DOMString elementNs = "", optional DOMString attrNs = "");
```
Can we special out Namespace instead of abbreviating it?

lukewarlow replied:
Seeing as it's a parameter name it's internal only so I'll update that to be fully spelt out.

rniwa commented on Source/WebCore/dom/TrustedTypePolicyFactory.idl:
```diff
@@ -0,0 +1,43 @@
+
+    DOMString? getAttributeType(DOMString tagName, DOMString attribute, optional DOMString elementNs = "", optional DOMString attrNs = "");
+
+    DOMString? getPropertyType(DOMString tagName, DOMString property, optional DOMString elementNs = "");
```
Ditto about ns.

lukewarlow committed:
Add stubs for Trusted Types API
https://bugs.webkit.org/show_bug.cgi?id=266980

Reviewed by Ryosuke Niwa and Alex Christensen.

This patch adds a runtime feature flag and the initial bindings for the Trusted Types API.
None of the enforcement logic is implemented yet, nor any changes to the DOM sinks.

Spec: https://w3c.github.io/trusted-types/dist/spec

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/content-security-policy/securitypolicyviolation/source-file-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/DOMParser-parseFromString-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/DOMWindowTimers-setTimeout-setInterval-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Document-write-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Element-insertAdjacentHTML-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Element-insertAdjacentText-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Element-outerHTML-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Element-setAttribute-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Element-setAttributeNS-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/GlobalEventHandlers-onclick-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLElement-generic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/HTMLScriptElement-internal-slot-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Node-multiple-arguments-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Range-createContextualFragment-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedType-AttributeNodes-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicy-CSP-no-name-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicy-CSP-wildcard-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicy-createXXX-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-blocking-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-constants-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-createXYZTests-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-cspTests-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-cspTests-noNamesGiven-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-cspTests-none-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-cspTests-none-skip-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-cspTests-wildcard-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-nameTests-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-createPolicy-unenforced-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-defaultPolicy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getAttributeType-namespace-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getPropertyType-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-isXXX-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-metadata-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/Window-TrustedTypes-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/WorkerGlobalScope-eval-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/WorkerGlobalScope-importScripts-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-Document-execCommand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-Node-multiple-arguments-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-DOMParser-parseFromString-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-DOMWindowTimers-setTimeout-setInterval-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Document-write-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Element-insertAdjacentHTML-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Element-outerHTML-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Element-setAttribute-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Element-setAttributeNS-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-HTMLElement-generic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-string-assignment-to-Range-createContextualFragment-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/block-text-node-insertion-into-script-element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/csp-block-eval-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/default-policy-callback-arguments-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/default-policy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/default-policy-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/empty-default-policy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/empty-default-policy-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-no-tt-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-tt-default-policy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-csp-tt-no-default-policy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-function-constructor-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-no-csp-no-tt-default-policy-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-no-csp-no-tt-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/eval-with-permissive-csp-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/no-require-trusted-types-for-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/no-require-trusted-types-for-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/require-trusted-types-for-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-createHTMLDocument-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-duplicate-names-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-duplicate-names-list-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-duplicate-names-list-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-duplicate-names-without-enforcement-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-eval-reporting-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-eval-reporting-no-unsafe-eval-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-eval-reporting-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-from-literal-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-report-only-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-reporting-check-report-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-source-file-path-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-svg-script-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-tojson-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/tt-block-eval-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/worker-constructor.https-expected.txt:
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/CreateHTMLCallback.h: Added.
* Source/WebCore/dom/CreateHTMLCallback.idl: Added.
* Source/WebCore/dom/CreateScriptCallback.h: Added.
* Source/WebCore/dom/CreateScriptCallback.idl: Added.
* Source/WebCore/dom/CreateScriptURLCallback.h: Added.
* Source/WebCore/dom/CreateScriptURLCallback.idl: Added.
* Source/WebCore/dom/TrustedHTML.cpp: Added.
(WebCore::TrustedHTML::create):
(WebCore::TrustedHTML::TrustedHTML):
* Source/WebCore/dom/TrustedHTML.h: Added.
(WebCore::TrustedHTML::toString const):
(WebCore::TrustedHTML::toJSON const):
* Source/WebCore/dom/TrustedHTML.idl: Added.
* Source/WebCore/dom/TrustedScript.cpp: Added.
(WebCore::TrustedScript::create):
(WebCore::TrustedScript::TrustedScript):
* Source/WebCore/dom/TrustedScript.h: Added.
* Source/WebCore/dom/TrustedScript.idl: Added.
* Source/WebCore/dom/TrustedScriptURL.cpp: Added.
(WebCore::TrustedScriptURL::create):
(WebCore::TrustedScriptURL::TrustedScriptURL):
* Source/WebCore/dom/TrustedScriptURL.h: Added.
(WebCore::TrustedScriptURL::toString const):
(WebCore::TrustedScriptURL::toJSON const):
* Source/WebCore/dom/TrustedScriptURL.idl: Added.
* Source/WebCore/dom/TrustedTypePolicy.cpp: Added.
(WebCore::TrustedTypePolicy::create):
(WebCore::TrustedTypePolicy::TrustedTypePolicy):
(WebCore::TrustedTypePolicy::createHTML):
(WebCore::TrustedTypePolicy::createScript):
(WebCore::TrustedTypePolicy::createScriptURL):
* Source/WebCore/dom/TrustedTypePolicy.h: Added.
(WebCore::TrustedTypePolicy::name const):
* Source/WebCore/dom/TrustedTypePolicy.idl: Added.
* Source/WebCore/dom/TrustedTypePolicyFactory.cpp: Added.
(WebCore::TrustedTypePolicyFactory::create):
(WebCore::TrustedTypePolicyFactory::createPolicy):
(WebCore::TrustedTypePolicyFactory::isHTML const):
(WebCore::TrustedTypePolicyFactory::isScript const):
(WebCore::TrustedTypePolicyFactory::isScriptURL const):
(WebCore::TrustedTypePolicyFactory::emptyHTML const):
(WebCore::TrustedTypePolicyFactory::emptyScript const):
(WebCore::TrustedTypePolicyFactory::getAttributeType const):
(WebCore::TrustedTypePolicyFactory::getPropertyType const):
* Source/WebCore/dom/TrustedTypePolicyFactory.h: Added.
(WebCore::TrustedTypePolicyFactory::defaultPolicy const):
* Source/WebCore/dom/TrustedTypePolicyFactory.idl: Added.
* Source/WebCore/dom/TrustedTypePolicyOptions.h: Added.
* Source/WebCore/dom/TrustedTypePolicyOptions.idl: Added.
* Source/WebCore/dom/WindowOrWorkerGlobalScope+TrustedTypes.idl: Added.
* Source/WebCore/dom/WindowOrWorkerGlobalScopeTrustedTypes.cpp: Added.
(WebCore::DOMWindowTrustedTypes::supplementName):
(WebCore::WorkerGlobalScopeTrustedTypes::supplementName):
(WebCore::DOMWindowTrustedTypes::DOMWindowTrustedTypes):
(WebCore::DOMWindowTrustedTypes::from):
(WebCore::DOMWindowTrustedTypes::trustedTypes const):
(WebCore::WorkerGlobalScopeTrustedTypes::WorkerGlobalScopeTrustedTypes):
(WebCore::WorkerGlobalScopeTrustedTypes::from):
(WebCore::WorkerGlobalScopeTrustedTypes::trustedTypes const):
(WebCore::WindowOrWorkerGlobalScopeTrustedTypes::trustedTypes):
* Source/WebCore/dom/WindowOrWorkerGlobalScopeTrustedTypes.h: Added.
* Source/WebCore/page/ios/DOMTimerHoldingTank.h:
* Source/WebCore/Modules/WebGPU/GPUOutOfMemoryError.h:

Canonical link: https://commits.webkit.org/273817@main
